### DISCUSSION TITLE
PHP 8 Compatibility

### DISCUSSION THREAD

**Discussion by AlexandruGG** [loophp/collection#103] [General]
Currently `loophp/collection` can be installed in a project running PHP 8, however does not have full compatibility with it. 

I started a branch to look into compatibility issues, and good news is the only one that seems to be problematic so far is the `asyncMap` operation.

I think what we need to do is:
1. enable CI runnning on both php-version 7.4 and 8.0 (easy)
2. fix any static analysis issues (easy)
3. decide what to do about `asyncMap` and implement plan (medium)

Regarding point 3, here are some thoughts:
- to use that operation requires [amphp/parallel-functions](https://github.com/amphp/parallel-functions), which is a bit like a "hidden" dependency. You only realise you need it at runtime because it will throw an exception, or by reading the [documentation](https://loophp-collection.readthedocs.io/en/latest/pages/api.html#asyncmap) of course
- the library it uses does not have PHP 8 support and there don't seem to be any plans to add it. Also, it doesn't look very actively maintained - latest release was in July 2020 
- is `asyncMap` that useful? Surely a nice to have but I imagine very few users actually benefit from it; and those actually using a lot of concurrency in PHP probably won't rely on it because they will already have their own things

This being said, I see a couple of options regarding `asyncMap`:
A) remove it from the next major version with PHP 8 support. It would still be available for everyone using the library on PHP 7.4
B) keep it in the library but "disable" it for PHP 8 users, i.e. throw a type of `LogicException` when used. This would be clearly specified in the documentation as well
C) add support and needed logic using a different library for parallelism, like https://github.com/spatie/fork. Note that his one is only available on PHP 8, so we'd have to check the running php version in the `AsyncMap` operation and use the corresponding library. On a quick search I couldn't find an equivalent package that supports both 7.4 and 8.0.

**Comment by AlexandruGG** [comment]
@drupol let me know your thoughts on this! 😁

**Comment by drupol** [comment]
Hi Alex,

We can slowly move on by fixing the 2 first points, then, for the last point, I can discuss with the maintainer (@kelunik) and see how we can update properly. 
If we can't get around it, I don't mind removing it in the meantime.

**Comment by kelunik** [comment]
What makes you think there's no PHP 8 support or no plans to add it? Having no release for a year can also mean it's simply stable.

**Reply by AlexandruGG** [reply]
hey @kelunik! 👋 


> What makes you think there's no PHP 8 support or no plans to add it?

my phrasing was "there don't _seem_ to be any plans to add it.", and it was based on what I can see publicly in [issues](https://github.com/amphp/parallel-functions/issues) and [milestones](https://github.com/amphp/parallel-functions/milestones). If there are plans to add it that's great 👍

> Having no release for a year can also mean it's simply stable.

that's true, it's a valid interpretation. I meant no offence with that, but it can be seen both ways I think 😅

**Reply by kelunik** [reply]
If something needs to be done to support PHP 8, please open an issue.

**Reply by AlexandruGG** [reply]
I was looking to follow your suggestion, but it looks like the failure only happens when running PHPSpec. Still not sure if the issue sits with PHPSpec, but here's what I'm seeing in this project:

`Error("Loop exceptionally stopped without resolving the promise")` coming from [wait](https://github.com/amphp/amp/blob/master/lib/functions.php#L231)

A deeper look shows `warning: stream_socket_accept(): Accept failed: Operation timed out` coming from [ProcessHub](https://github.com/amphp/parallel/blob/master/lib/Context/Internal/ProcessHub.php#L78)

Any clue what could be causing this timeout? @kelunik 

I tried to also run the PHPUnit tests in [amphp/parallel-functions](https://github.com/amphp/parallel-functions) under PHP 8 (just trying to find some clues 😄), however things are failing there because the PHPUnit version doesn't support PHP 8; it would need to be updated to 9.3 for that because of [this issue](https://github.com/sebastianbergmann/phpunit/issues/4373) regarding `match`.

**Reply by kelunik** [reply]
The suppressed warning is normal I guess, as there's a while to accept all immediately acceptable connections.

What's the previous exception of the error thrown from wait?

**Reply by AlexandruGG** [reply]
@kelunik after doing some more debugging, here's what I found so far:

- there is no _previous_ in the throwable in `wait` (it's NULL)
- in PHP 8 the warning suppressed in ProcessHub is an exception. I added a try-catch around the [while loop here](https://github.com/amphp/parallel/blob/master/lib/Context/Internal/ProcessHub.php#L78) and dumped the class and trace. Here it is:
```
string(40) "PhpSpec\Exception\Example\ErrorException"
string(170) "warning: stream_socket_accept(): Accept failed: Operation timed out in /Users/alex/dev/forked/collection/vendor/amphp/parallel/lib/Context/Internal/ProcessHub.php line 81"
string(5959) "#0 [internal function]: PhpSpec\Runner\Maintainer\ErrorMaintainer->errorHandler(2, 'stream_socket_a...', '/Users/alex/dev...', 81)
#1 /Users/alex/dev/forked/collection/vendor/amphp/parallel/lib/Context/Internal/ProcessHub.php(81): stream_socket_accept(Resource id #979, 0)
#2 /Users/alex/dev/forked/collection/vendor/amphp/amp/lib/Loop/NativeDriver.php(321): Amp\Parallel\Context\Internal\ProcessHub::Amp\Parallel\Context\Internal\{closure}('a', Resource id #979, NULL)
#3 /Users/alex/dev/forked/collection/vendor/amphp/amp/lib/Loop/NativeDriver.php(127): Amp\Loop\NativeDriver->selectStreams(Array, Array, 4.996)
#4 /Users/alex/dev/forked/collection/vendor/amphp/amp/lib/Loop/Driver.php(138): Amp\Loop\NativeDriver->dispatch(true)
#5 /Users/alex/dev/forked/collection/vendor/amphp/amp/lib/Loop/Driver.php(72): Amp\Loop\Driver->tick()
#6 /Users/alex/dev/forked/collection/vendor/amphp/amp/lib/Loop.php(95): Amp\Loop\Driver->run()
#7 /Users/alex/dev/forked/collection/vendor/amphp/amp/lib/functions.php(229): Amp\Loop::run(Object(Closure))
#8 /Users/alex/dev/forked/collection/src/Operation/AsyncMap.php(93): Amp\Promise\wait(Object(Amp\Internal\PrivatePromise))
#9 /Users/alex/dev/forked/collection/src/Iterator/ClosureIterator.php(55): loophp\collection\Operation\AsyncMap::loophp\collection\Operation\{closure}(Object(loophp\collection\Iterator\ClosureIterator))
#10 [internal function]: loophp\collection\Iterator\ClosureIterator->getGenerator()
#11 /Users/alex/dev/forked/collection/src/Iterator/ProxyIterator.php(66): Generator->valid()
#12 /Users/alex/dev/forked/collection/vendor/phpspec/phpspec/src/PhpSpec/Matcher/Iterate/IterablesMatcher.php(42): loophp\collection\Iterator\ProxyIterator->valid()
#13 /Users/alex/dev/forked/collection/vendor/phpspec/phpspec/src/PhpSpec/Matcher/IterateAsMatcher.php(52): PhpSpec\Matcher\Iterate\IterablesMatcher->match(Object(loophp\collection\Collection), Array)
#14 /Users/alex/dev/forked/collection/vendor/phpspec/phpspec/src/PhpSpec/Wrapper/Subject/Expectation/Positive.php(34): PhpSpec\Matcher\IterateAsMatcher->positiveMatch('iterateAs', Object(loophp\collection\Collection), Array)
#15 /Users/alex/dev/forked/collection/vendor/phpspec/phpspec/src/PhpSpec/Wrapper/Subject/Expectation/DispatcherDecorator.php(68): PhpSpec\Wrapper\Subject\Expectation\Positive->match('iterateAs', Object(loophp\collection\Collection), Array)
#16 /Users/alex/dev/forked/collection/vendor/phpspec/phpspec/src/PhpSpec/Wrapper/Subject/Expectation/UnwrapDecorator.php(37): PhpSpec\Wrapper\Subject\Expectation\DispatcherDecorator->match('iterateAs', Object(loophp\collection\Collection), Array)
#17 /Users/alex/dev/forked/collection/vendor/phpspec/phpspec/src/PhpSpec/Wrapper/Subject/Expectation/ConstructorDecorator.php(51): PhpSpec\Wrapper\Subject\Expectation\UnwrapDecorator->match('iterateAs', Object(loophp\collection\Collection), Array)
#18 /Users/alex/dev/forked/collection/vendor/phpspec/phpspec/src/PhpSpec/Wrapper/Subject.php(291): PhpSpec\Wrapper\Subject\Expectation\ConstructorDecorator->match('iterateAs', Object(PhpSpec\Wrapper\Subject), Array, Object(PhpSpec\Wrapper\Subject\WrappedObject))
#19 /Users/alex/dev/forked/collection/vendor/phpspec/phpspec/src/PhpSpec/Wrapper/Subject.php(240): PhpSpec\Wrapper\Subject->callExpectation('shouldIterateAs', Array)
#20 /Users/alex/dev/forked/collection/spec/loophp/collection/CollectionSpec.php(241): PhpSpec\Wrapper\Subject->__call('shouldIterateAs', Array)
#21 [internal function]: spec\loophp\collection\CollectionSpec->it_can_asyncMap()
#22 /Users/alex/dev/forked/collection/vendor/phpspec/phpspec/src/PhpSpec/Runner/ExampleRunner.php(143): ReflectionMethod->invokeArgs(Object(spec\loophp\collection\CollectionSpec), Array)
#23 /Users/alex/dev/forked/collection/vendor/phpspec/phpspec/src/PhpSpec/Runner/ExampleRunner.php(77): PhpSpec\Runner\ExampleRunner->executeExample(Object(spec\loophp\collection\CollectionSpec), Object(PhpSpec\Loader\Node\ExampleNode))
#24 /Users/alex/dev/forked/collection/vendor/phpspec/phpspec/src/PhpSpec/Runner/SpecificationRunner.php(57): PhpSpec\Runner\ExampleRunner->run(Object(PhpSpec\Loader\Node\ExampleNode))
#25 /Users/alex/dev/forked/collection/vendor/phpspec/phpspec/src/PhpSpec/Runner/SuiteRunner.php(52): PhpSpec\Runner\SpecificationRunner->run(Object(PhpSpec\Loader\Node\SpecificationNode))
#26 /Users/alex/dev/forked/collection/vendor/phpspec/phpspec/src/PhpSpec/Console/Command/RunCommand.php(173): PhpSpec\Runner\SuiteRunner->run(Object(PhpSpec\Loader\Suite))
#27 /Users/alex/dev/forked/collection/vendor/symfony/console/Command/Command.php(256): PhpSpec\Console\Command\RunCommand->execute(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#28 /Users/alex/dev/forked/collection/vendor/symfony/console/Application.php(989): Symfony\Component\Console\Command\Command->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#29 /Users/alex/dev/forked/collection/vendor/symfony/console/Application.php(290): Symfony\Component\Console\Application->doRunCommand(Object(PhpSpec\Console\Command\RunCommand), Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#30 /Users/alex/dev/forked/collection/vendor/phpspec/phpspec/src/PhpSpec/Console/Application.php(87): Symfony\Component\Console\Application->doRun(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#31 /Users/alex/dev/forked/collection/vendor/symfony/console/Application.php(166): PhpSpec\Console\Application->doRun(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#32 /Users/alex/dev/forked/collection/vendor/phpspec/phpspec/bin/phpspec(25): Symfony\Component\Console\Application->run()
#33 /Users/alex/dev/forked/collection/vendor/phpspec/phpspec/bin/phpspec(27): {closure}('7.0.1')
#34 {main}"
```

After some more digging in the PHPSpec code I found the culprit 🎉 . [Here in ErrorMaintainer](https://github.com/phpspec/phpspec/blob/main/src/PhpSpec/Runner/Maintainer/ErrorMaintainer.php#L95), in PHP 8 the function `error_reporting` [returns a non-zero code for warnings](https://www.php.net/manual/en/function.error-reporting.php#125674). I tested and in PHP 7.4 it returns 0 during the tests for `asyncMap`, but with PHP 8 it returns `4436` for me.

So I think we can chalk this down to a potential issue with PHPSpec, right? It's now treating E_WARNINGs like errors, which unless it was an intentional decision is a valid issue.

**Reply by drupol** [reply]
Hi!

`4436` is `1000101010100` in binary and correspond to these:

- `E_PARSE` 
- `E_CORE_ERROR` 
- `E_COMPILE_ERROR` 
- `E_USER_ERROR` 
- `E_RECOVERABLE_ERROR`

I see that the behavior of `\error_reporting()` has changed in PHP 8, maybe we should open a PR on [phpspec](https://github.com/phpspec/phpspec) project.

Ping @ciaranmcnulty : Hi mate! Do you have something to say here? Do you think we could fix that in PHPSpec?

**Reply by drupol** [reply]
By doing:

```diff
diff --git a/src/PhpSpec/Runner/Maintainer/ErrorMaintainer.php b/src/PhpSpec/Runner/Maintainer/ErrorMaintainer.php
index b8f5d0ea..3b3c177d 100644
--- a/src/PhpSpec/Runner/Maintainer/ErrorMaintainer.php
+++ b/src/PhpSpec/Runner/Maintainer/ErrorMaintainer.php
@@ -117,7 +117,9 @@ final class ErrorMaintainer implements Maintainer
             }
         }

-        if (0 !== error_reporting()) {
+        $errorLevel = E_PARSE | E_CORE_ERROR | E_COMPILE_ERROR | E_USER_ERROR | E_RECOVERABLE_ERROR;
+
+        if ($errorLevel !== error_reporting()) {
             throw new ExampleException\ErrorException($level, $message, $file, $line);
         }
```

It seems to fix the issue.

However, it would be really nice to have some more inputs.

**Reply by AlexandruGG** [reply]
I guess this thing is related to the default error reporting level in PHP 8, which was changed to E_ALL: https://php.watch/versions/8.0/error-display-E_ALL

**Reply by drupol** [reply]
Oh nice!

I think we need to address that issue to PHPSpec, let's wait for @ciaranmcnulty's input.

**Reply by kelunik** [reply]
The proper check is `if (!(error_reporting() & $errno)) {`

**Reply by drupol** [reply]
Aaah true, we must "bitmask" the `error_reporting()` return value !

Nice nice nice!

**Reply by ciaranmcnulty** [reply]
IIRC @pamil had a similar issue with Sylius?

**Reply by ciaranmcnulty** [reply]
See https://github.com/Sylius/Sylius/blob/master/phpspec.yml.dist#L46

**Reply by ciaranmcnulty** [reply]
(Maybe this should be a PhpSpec docs issue?)

**Reply by drupol** [reply]
Hi Ciaran,

I tried, I end up with the same original issue.

**Reply by AlexandruGG** [reply]
by default I see PHPSpec uses `30719` which represents `E_ALL`

@drupol for our issue I tested and it works with `22525 # E_ALL & ~E_DEPRECATED & ~E_STRICT & ~E_WARNING`

**Reply by drupol** [reply]
I think it's definitely an issue with PHPSpec.
Look, I'm currently working for work and I'm having the same kind of issue, see:

![image](https://user-images.githubusercontent.com/252042/120457489-c9052380-c396-11eb-9a88-320c2f6f06c9.png)

The test can be found here: https://github.com/ecphp/cas-bundle/blob/master/spec/EcPhp/CasBundle/Security/Core/User/CasUserProviderSpec.php#L72

**Reply by ciaranmcnulty** [reply]
Can we move this to a phpspec issue?

**Reply by drupol** [reply]
Done! https://github.com/phpspec/phpspec/issues/1375

**Reply by AlexandruGG** [reply]
https://github.com/phpspec/phpspec/pull/1378 was merged so this is all good now. Thanks for the assistance @kelunik! 🎉
